pkg/consensus/consensus.go

```diff
@@ -216,35 +216,49 @@ func (s *service) eventLoop() {
 
 func (s *service) newPayload() payload.ConsensusPayload {
 	return &Payload{
-		message: new(message),
+		message: &message{
+			stateRootEnabled: s.stateRootEnabled(),
+		},
 	}
 }
 
+// stateRootEnabled checks if state root feature is enabled on current height.
+// It should be called only from dbft callbacks and is not protected by any mutex.
+func (s *service) stateRootEnabled() bool {
+	return s.Chain.GetConfig().EnableStateRoot
+}
+
 func (s *service) newPrepareRequest() payload.PrepareRequest {
-	sr, err := s.Chain.GetStateRoot(s.Chain.BlockHeight())
-	if err != nil {
+	if !s.stateRootEnabled() {
 		return new(prepareRequest)
 	}
-	return &prepareRequest{
-		proposalStateRoot: sr.MPTRootBase,
+	sr, err := s.Chain.GetStateRoot(s.Chain.BlockHeight())
+	if err == nil {
+		return &prepareRequest{
+			stateRootEnabled:  s.stateRootEnabled(),
+			proposalStateRoot: sr.MPTRootBase,
+		}
 	}
+	return &prepareRequest{stateRootEnabled: s.stateRootEnabled()}
 }
 
 func (s *service) newCommit() payload.Commit {
+	if !s.stateRootEnabled() {
+		return new(commit)
+	}
+	c := &commit{stateRootEnabled: s.stateRootEnabled()}
```

Comment:
You already know the value of `stateRootEnabled` here. `s.stateRootEnabled()` involves getting a config from the chain (copying it in the process), it's an obvious overhead that's easy to avoid.

Reply:
Fixed.